src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,627 @@
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Direction {
+  string name = 1;        // Name of the directional argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Offset argument of a StencilFunction
+//
+// Note that this message is merely a placeholder as the actual arguments is encoded as an AST node.
+// @ingroup sir_proto
+message Offset {
+  string name = 1;        // Name of the offset argument
+  SourceLocation loc = 2; // Source location
+}
+
+// @brief Directional argument of a StencilFunction
+// @ingroup sir_proto
+message StencilFunctionArg {
+  // One of Field, Direction or Offset
+  oneof Arg {
+    Field field_value = 1;
+    Direction direction_value = 2;
+    Offset offset_value = 3;
+  }
+}
+
+// @brief Representation of a vertical interval, given by a lower and upper bound where a bound
+// is represented by a level and an offset (`bound = level + offset`)
+//
+// The Interval has to satisfy the following invariants:
+//  - `lower_level >= Interval::Start`
+//  - `upper_level <= Interval::End`
+//  - `(lower_level + lower_offset) <= (upper_level + upper_offset)`
+//
+// @ingroup sir_proto
+message Interval {
+  // Indication of a special level
+  enum SpecialLevel {
+    Start = 0; // The first level
+    End = 1;   // The last level
+  }
+
+  // Lower level
+  //
+  // A level can be any integer between `[Start, End]` while the `Start` and `End` levels are
+  // treated specially.
+  oneof LowerLevel {
+    SpecialLevel special_lower_level = 1;
+    int32 lower_level = 2;
+  }
+
+  // Upper level
+  //
+  // A level can be any integer between `[Start, End]` while the `Start` and `End` levels are
+  // treated specially.
+  oneof UpperLevel {
+    SpecialLevel special_upper_level = 3;
+    int32 upper_level = 4;
+  }
+
+  int32 lower_offset = 5; // Lower offset added to `lower_level`
+  int32 upper_offset = 6; // Upper offset added to `upper_level`
+}
+
+// @brief Supported builtin types
+// @ingroup sir_proto
+message BuiltinType {
+  enum TypeID { None = 0; Auto = 1; Boolean = 2; Integer = 3; Float = 4; }
+
+  TypeID type_id = 1; // Type identifier
+}
+
+// @brief Supported builtin types
+// @ingroup sir_proto
+message Dimension {
+  enum Direction { I = 0; J = 1; K = 2; Invalid = 3; }
+
+  Direction dimension = 1; // Dimension identifier
+}
+
+// @brief Supported builtin types
+// @ingroup sir_proto
+message Type {
+  // Underlying type (either a custom type given by `name` or a `BuiltinType`)
+  oneof type {
+    string name = 1;              // Name of the custom type
+    BuiltinType builtin_type = 2; // ID of the builtin type
+  }
+
+  bool is_const = 3;    // Is the type const qualified?
+  bool is_volatile = 4; // Is the type volatile qualified?
+}
+
+// @brief Declaration of a vertical region
+// @ingroup sir_proto
+message VerticalRegion {
+  enum LoopOrder { Forward = 0; Backward = 1; }
```

Comment:
do we need an "any" as well?

Reply:
The way the MultiStage splitting passes work is that these loop orders serve as the fallback value of the loop order if a parallel loop order is not possible, meaning we *always* assume a parallel loop order and adjust it to the user provided one if necessary.

Reply:
yes, it does not make sense in dawn, but as a user description of the stencil, I wouldnt know what to generate for a horizontal stencil, having to specify forward or backward is ambiguous. So I would still add the keyword in the SIR, and simply not use it internally by gtclang and dawn

Reply:
As discussed I will integrate it.